### PR TITLE
feat: check that locality ID match with resource locality

### LINKED ISSUE #1673: provider: check locality IDs
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

Most resources export a locality ID (`fr-par-1/xxxx-xxxx`), when linking multiple resources altogether we should check that the localities match when needed.

For example, an instance should check that the ip zoned received is in the same zone.

### Terraform snippet

```hcl

resource scaleway_instance_ip ip {
  zone = "fr-par-2"
}

resource scaleway_instance_server server {
  type = "DEV1-S"
  zone = "fr-par-1"
  image = "ubuntu_jammy"
  ip_id = scaleway_instance_ip.ip.id
}
```

This snippet should fail on `terraform plan`

### New or Affected Resource(s)

A lot

### References

* #1671

### PR DESCRIPTION
**Pull request by Codelax**
Closes #1673

### CODE REVIEW

File: scaleway/resource_instance_server.go [hunk 1]
```diff
@@ -270,6 +270,11 @@ func resourceScalewayInstanceServer() *schema.Resource {
 			"organization_id": organizationIDSchema(),
 			"project_id":      projectIDSchema(),
 		},
+		CustomizeDiff: customizeDiffLocalityCheck(
```
Comment (remyleone) [inline scaleway/resource_instance_server.go:273]: Having a sequence would make it easier to read here. https://pkg.go.dev/github.com/hashicorp/terraform-plugin-sdk/helper/customdiff#Sequence

File: scaleway/resource_baremetal_server.go [hunk 2]
```diff
@@ -219,7 +220,7 @@ If this behaviour is wanted, please set 'reinstall_on_ssh_key_changes' argument
 				},
 			},
 		},
-		CustomizeDiff: func(_ context.Context, diff *schema.ResourceDiff, i interface{}) error {
+		CustomizeDiff: func(ctx context.Context, diff *schema.ResourceDiff, i interface{}) error {
```
Comment (remyleone) [inline scaleway/resource_baremetal_server.go:223]: Could you extract this function in a named function? And combine it with a sequence ?

File: scaleway/helpers.go [hunk 3]
```diff
@@ -848,29 +848,87 @@ func retryOnTransientStateError[T any, U any](action func() (T, error), waiter f
 	return t, err
 }
 
+// compareLocalities compare two localities
+// They are equal if they are the same or if one is a zone contained in the other
```
Comment (remyleone) [inline scaleway/helpers.go:852]: I'm not sure to understand this sentence. What does it mean for a zone to be contained in another zone? Are you talking about region and zone comparison?

Comment (Codelax) [reply scaleway/helpers.go]: Comment have been improved

File: scaleway/resource_container_domain.go [hunk 4]
```diff
@@ -47,6 +47,7 @@ func resourceScalewayContainerDomain() *schema.Resource {
 			},
 			"region": regionSchema(),
 		},
+		CustomizeDiff: customizeDiffLocalityCheck("container_id"),
```
[the remark this answers is not in the document]
Comment (Codelax) [reply scaleway/resource_container_domain.go:50]: Make sense when using multiple customDiff function, but this is rare

**Review by remyleone — APPROVED** [review]

**Pull request merged**